Expose custom attributes from C++ functions

lantiga opened the pull request:
This PR demonstrates a possible way to expose custom attributes from C++ functions.
Motivation: right now functions implemented in C++ do not expose function-specific attributes or methods (e.g. `stride` for `ConvForward`). This makes it impossible to get such attributes while traversing a computation graph.

Currently only convolution attributes have been exposed, with the aim of illustrating the approach. If the proposal gets a thumbs up, I'll proceed and expose attributes for the rest of the C++ functions.

As agreed with @apaszke on Slack #autograd-internals, this PR points to the `autograd` branch.

lantiga committed:
Added custom properties and methods to functions (conv for now).

lantiga committed:
Add missing include to fix macOS build

lantiga committed:
Replace PyInt_FromLong with PyLong_FromLong to fix Python 3 build

lantiga committed:
Replace PyInt_FromLong with PyLong_FromLong to fix Python 3 build

albanD commented on torch/csrc/autograd/python_cpp_function.cpp:
```diff
@@ -88,67 +89,17 @@ void THPCppFunction_dealloc(PyObject* self)
   type.tp_call = THPCppFunction_call;
-  type.tp_methods = THPCppFunction_methods;
-  type.tp_getset = THPCppFunction_properties;
+  type.tp_methods = function_methods ? function_methods : attributes::default_methods;
```
Would it be possible here to add the default methods even when function_methods is provided? That would allow to change the common methods without modifying every function attribute and reduce the chances of forgetting one of the default one?
(same for the properties below)

lantiga replied:
I agree. We need to build a static array in stages, I'm thinking about a macro adding the default entries. What do you think?

albanD replied:
I would do it with a macro, but maybe there is a nice cpp way to do this @apaszke ?

lantiga replied:
Done with macro in 7b64a8a1260e8885bdb749fa1b5421a3cf4268be.

albanD commented on torch/csrc/autograd/functions/convolution_attr.h:
```diff
@@ -0,0 +1,105 @@
+namespace attributes {
+
+template<class T>
+PyObject* conv_stride(THPCppFunction* self, PyObject* hook)
```
I am not sure what are the guidelines for cpp code but these getter functions look quite similar.
Could we do a macro (or the correct cpp way to do this) like `IMPLEMENT_GET_VECTOR(conv_stride, ConvForward, stride)` or `IMPLEMENT_GET_NUMBER(conv_groups, ConvForward, groups)` that could be reused for all functions?

apaszke replied:
I agree. I'd be the best to implement a number of helper templates/macros and use these to implement attributes for all new functions in a few lines. See e.g. [some helpers from `python_function.cpp`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_function.cpp#L715-L748). It should be possible to implement similar ones for each type (e.g. getIntMember, getBoolMember, etc.).

apaszke replied:
Also, once the definitions are smaller they don't need to be in an additional file.

lantiga replied:
Agreed

lantiga replied:
constexpr and std::tuple_cat I would guess

lantiga replied:
Done in 932ac3a981b4a8301e05f767872905f6d78991fe.

albanD reviewed (commented):
I am not sure how big this `functions` folder is going to grow. We may want to separate the attributes code in a separate sub-folder to keep this one with just the actual function implementations?

apaszke commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -291,7 +291,7 @@ int THPVariable_set_volatile(THPVariable *self, PyObject *obj)
 {
   auto& var = *self->cdata;
-  return PyInt_FromLong(var.output_nr);
+  return PyLong_FromLong(var.output_nr);
```
I think it's better to leave it as `PyInt`. It's macroed to `PyLong` in Python 3, and is an `int` object in Python 2.

lantiga replied:
Ok, I'll revert the commit

lantiga replied:
`THP.h`, didn't notice it at first. Reverting the related commits.

lantiga replied:
Reverted a8857be5b5ff5cf73de04b8821248aef9ae6b703 and f34595362cce809f15da1e8e558153e3d946835b. Included `THP.h` in fe8b1524cc4fdd051ddecb7172849b1a28ba0ff0.

apaszke commented on torch/csrc/autograd/python_cpp_function_attr.cpp:
```diff
@@ -0,0 +1,54 @@
+
+namespace attributes {
+
+PyObject* next_functions(THPCppFunction* self, PyObject* hook)
```
I think it's better to keep these in `python_cpp_function.cpp` and try to not add more files.

lantiga replied:
For convolution_attr.h: I wanted to avoid polluting convolution.h and convolution.cpp with Python API stuff. Are you thinking about having these definitions in init.cpp directly?

lantiga replied:
Done in 932ac3a981b4a8301e05f767872905f6d78991fe.

lantiga committed:
Revert "Replace PyInt_FromLong with PyLong_FromLong to fix Python 3 build"

This reverts commit a9a35ffa14de9138106bb48a9561b0f7ab83ac81.

lantiga committed:
Revert "Replace PyInt_FromLong with PyLong_FromLong to fix Python 3 build"

This reverts commit cef68e8e279e68ec2e44692c2af635af9375b1b0.

lantiga committed:
Include THP for PyInt macros

lantiga committed:
Add helpers to avoid code repetition for attributes. Remove attr files.

lantiga commented:
Ok. I have implemented a `getTupleAttr` and `getValueAttr`, to avoid code repetition for exposing attributes. I've also removed `_attr.*` files, as a result of the code shrink. For the same reason I got rid of the `attributes` namespace.

The one thing I still have to do is to inherit default attributes.

lantiga committed:
Include THP for PyInt_ functions in Python 3

lantiga committed:
Fix formatting and typo

lantiga commented:
Likely unrelated: for some reason one of the dataloader tests failed on Travis CI for Python 3.5
```
Running dataloader tests
.....s...s....s.sFs...
======================================================================
FAIL: test_sequential_batch (__main__.TestDictDataLoader)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_dataloader.py", line 252, in test_sequential_batch
    self.assertEqual(set(sample.keys()), {'a_tensor', 'another_dict'})
  File "/home/travis/build/pytorch/pytorch/test/common.py", line 189, in assertEqual
    self.assertEqual(x_, y_, prec, message)
  File "/home/travis/build/pytorch/pytorch/test/common.py", line 186, in assertEqual
    super(TestCase, self).assertEqual(x, y)
AssertionError: 'a_tensor' != 'another_dict'
- a_tensor
+ another_dict
```

lantiga committed:
Introduce macros to avoid including default props and methods by hand

lantiga commented:
I played with a few C++ possibilities for concatenating `default_properties` to the ones passed to `_initFunctionPyTypeObject`, but the macro solution felt cleaner. I'll be happy to see C++ alternatives though.

At this point, all changes that were suggested are in.

lantiga commented:
New autograd merged, great!
Should I proceed rebasing against master and send a new PR? /cc @soumith @apaszke

soumith commented:
yes please. rebase against master, send a PR there.
